vision_evaluation/retrieval_evaluators.py

```diff
@@ -0,0 +1,156 @@
+import numpy as np
+import sklearn.metrics as sm
+
+from .evaluators import MemorizingEverythingEvaluator, PrecisionRecallCurveMixin
+from vision_evaluation.prediction_filters import TopKPredictionFilter
+
+
+class RetrievalEvaluator(MemorizingEverythingEvaluator):
+    def add_predictions(self, predictions, targets):
+        """ Add a batch of predictions for evaluation.
+        Args:
+            predictions: the model output array. Shape (N, M) where N is the number of queries and M is the number of items in the dataset
+            targets: the ground truths. Shape (N, M) where N is the number of queries and M is the number of items in the dataset
+        """
+        super(RetrievalEvaluator, self).add_predictions(predictions, targets)
+
+    def get_report(self):
+        return {self._get_id(): self.calculate_score()}
+
+
+class PrecisionAtKEvaluator(RetrievalEvaluator):
+    """
+    Precision evaluator for both multi-class and multi-label classification
+    """
+
+    def __init__(self, k):
+        super(PrecisionAtKEvaluator, self).__init__(TopKPredictionFilter(k))
+        self.k = k
+
+    def _get_id(self):
+        return f'precision_at_{self.k}'
+
+    def calculate_score(self):
```

Comment:
This might have potential issue, where if calculate_score is called in based class (as it is from base class), with args provided, it might fail I suppose.

Might be safer to directly have each class's overriding get_report specifying the right average and filter_out_zero_tgt